crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,849 @@
+    let obs = SparseObservable::zero(num_qubits);
+    Box::into_raw(Box::new(obs))
+}
+
+/// @ingroup QkSparseObservable
+/// Construct the identity observable.
+///
+/// @param num_qubits The number of qubits the observable is defined on.
+///
+/// @return A pointer to the created observable.
+///
+/// # Example
+///
+///     QkSparseObservable *identity = qk_obs_identity(100);
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_obs_identity(num_qubits: u32) -> *mut SparseObservable {
+    let obs = SparseObservable::identity(num_qubits);
+    Box::into_raw(Box::new(obs))
+}
+
+/// @ingroup QkSparseObservable
+/// Construct a new observable from raw data.
+///
+/// @param num_qubits The number of qubits the observable is defined on.
+/// @param num_terms The number of terms.
+/// @param num_bits The total number of non-identity bit terms.
+/// @param coeffs A pointer to the first element of the coefficients array, which has length
+///     ``num_terms``.
+/// @param bit_terms A pointer to the first element of the bit terms array, which has length
+///     ``num_bits``.
+/// @param indices A pointer to the first element of the indices array, which has length
+///     ``num_bits``. Note that, per term, these *must* be sorted incrementally.
+/// @param boundaries A pointer to the first element of the boundaries array, which has length
+///     ``num_terms + 1``.
+///
+/// @return If the input data was coherent and the construction successful, the result is a pointer
+///     to the observable. Otherwise a null pointer is returned.
+///
+/// # Example
+///
+///     // define the raw data for the 100-qubit observable |01><01|_{0, 1} - |+-><+-|_{98, 99}
+///     uint32_t num_qubits = 100;
+///     uint64_t num_terms = 2;  // we have 2 terms: |01><01|, -1 * |+-><+-|
+///     uint64_t num_bits = 4; // we have 4 non-identity bits: 0, 1, +, -
+///
+///     complex double coeffs[2] = {1, -1};
+///     QkBitTerm bits[4] = {QkBitTerm_Zero, QkBitTerm_One, QkBitTerm_Plus, QkBitTerm_Minus};
+///     uint32_t indices[4] = {0, 1, 98, 99};  // <-- e.g. {1, 0, 99, 98} would be invalid
+///     size_t boundaries[3] = {0, 2, 4};
+///
+///     QkSparseObservable *obs = qk_obs_new(
+///         num_qubits, num_terms, num_bits, coeffs, bits, indices, boundaries
+///     );
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following conditions are violated:
+///
+///   * ``coeffs`` is a pointer to a ``complex double`` array of length ``num_terms``
+///   * ``bit_terms`` is a pointer to a ``QkBitTerm`` array of length ``num_bits``
+///   * ``indices`` is a pointer to a ``uint32_t`` array of length ``num_bits``, which is
+///     term-wise sorted in strict ascending order, and every element is smaller than ``num_qubits``
+///   * ``boundaries`` is a pointer to a ``size_t`` array of length ``num_terms + 1``, which is
+///     sorted in ascending order, the first element is 0 and the last element is
+///     smaller than ``num_terms``
```

Comment:
It would be good to make clear in here that all the items in `bit_terms` _must_ be valid `BitTerm`s, since in C land, the type is just an alias to `uint8_t`, so tooling won't help.

Reply:
Added in 624c3dd